server/src/main/java/com/saralein/server/middleware/AuthMiddleware.java

```diff
@@ -0,0 +1,63 @@
+package com.saralein.server.middleware;
+
+import com.saralein.server.request.Request;
+import com.saralein.server.response.ErrorResponse;
+import com.saralein.server.response.Response;
+import com.saralein.server.router.Routes;
+import java.util.Base64;
+
+public class AuthMiddleware implements Middleware {
+    private final Routes routes;
+    private final String username;
+    private final String password;
+    private final String realm;
+    private Callable next;
+
+    public AuthMiddleware(Routes routes, String username, String password, String realm) {
+        this.routes = routes;
+        this.username = username;
+        this.password = password;
+        this.realm = realm;
+        this.next = null;
+    }
+
+    @Override
+    public Middleware apply(Callable callable) {
+        next = callable;
+        return this;
+    }
+
+    @Override
+    public Response call(Request request) {
+        if (isAuthorized(request)) {
+            return next.call(request);
+        }
+
+        return unauthorized();
+    }
+
+    private boolean isAuthorized(Request request) {
+        String uri = request.getUri();
+        String encodedAuthorization = request.getHeader("Authorization");
+        String parsedAuthorization = parseRequestAuthorization(encodedAuthorization);
+        return !routes.requiresAuthorization(uri) || encodeValidAuthorization().equals(parsedAuthorization);
+    }
```

Comment:
I think this is 90% of the way there, but I think this part feels like domain leak. I think it's strange for the routes to know about authorization. FWIW I don't think there's anything weird about the middleware getting a list of routes that need authorization, rather than all routes.

Reply:
Also, since this is a callable, and the controller is a callable, could you apply the middleware directly to the controller inside when you add it to the routes? 

That way, it would still be middleware, but it would be applied only for that controller.

Reply:
OK, I can adjust.  I remember having a list of routes a while ago, and I think there was a suggestion about somehow using the routes.  I could have also misconstrued the comment.  I seem to do that too frequently.

Reply:
We can do either/or. I just wanted to point out that option.

Reply:
I didn't make the controllers callable, but I could easily do that and wish I had thought of it. 
 Might fix some of these issues and then do it on another branch just for all the files it will change.